dhis2/dhis2-android-capture-app
# Initial proof of concept Idling Resources

## Description
The idea of this ticket is to add support for idling resources and coroutines. The core functionality consists of using a common dispatcher that is provided through DI. 

```kotlin
interface DispatcherProvider : CoroutineContext.Element {

  override val key: CoroutineContext.Key<*> get() = Key

  val default: CoroutineDispatcher
  val io: CoroutineDispatcher
  val main: CoroutineDispatcher
  val mainImmediate: CoroutineDispatcher
  val unconfined: CoroutineDispatcher

  companion object Key : CoroutineContext.Key<DispatcherProvider>
}

```

When it comes to testing (espresso) each dispatcher of the provider (IO, Default, Unconfined, Main) should be wrapped with **IdlingDispatcher**, which contains a counting Idling resource and waits for the coroutine to finish.

```kotlin
class IdlingDispatcher(
  private val delegate: CoroutineDispatcher
) : CoroutineDispatcher() {

  val counter: CountingIdlingResource = CountingIdlingResource("IdlingResource for $this")

  fun isIdle(): Boolean = counter.isIdleNow

  override fun dispatch(context: CoroutineContext, block: Runnable) {

    val runnable = Runnable {
      counter.increment()
      try {
        block.run()
      } finally {
        counter.decrement()
      }
    }
    delegate.dispatch(context, runnable)
  }

  override fun toString(): String = "CountingDispatcher delegating to $delegate"
}
```
After the dispatcher is wrapped with IdlingDispatcher we need to register before the tests.

```kotlin
// Fetch common APP dispatcher
val customDispatcherProvider = context.applicationContext.app().appComponent().customDispatcherProvider()

// Wrap it with IdlingDispatcherProvider and then use a rule to register the new idling and unregister
  @JvmField
  @Rule
  val idlingRule = IdlingDispatcherProviderRule {
      IdlingDispatcherProvider(customDispatcherProvider)
  }
```

```kotlin
class IdlingDispatcherProviderRule(
  private val factory: () -> IdlingDispatcherProvider
) : TestWatcher() {

  lateinit var dispatcherProvider: IdlingDispatcherProvider

  override fun starting(description: Description?) {
    dispatcherProvider = factory.invoke()

    dispatcherProvider.registerAllIdlingResources()
  }

  override fun finished(description: Description?) {
    dispatcherProvider.unregisterAllIdlingResources()
  }
}
```
With all of these one way to use the dispatchers through the app would be something like this: 
1. Create a coroutine Scope with the common dispatcher injected in the view model
2. Make sure you cancel it during onDestroy/Stop to avoid leaks

```kotlin
someCoroutineScope = CoroutineScope(
    Job() + main + dispatcher
)
```

Pending Tasks:
1) Remove all hardcoded dispatchers
2) Inject them
3) Move to common and fix dependencies
4) Try to use dagger in the form module

Notes: I had to undo some refactors regarding injecting and form module. Lets chat about this @Balcan @ferdyrod
